File: zlzlelzl/leetcode/Dynamic Programming/Coin Change.py
import sys


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [sys.maxsize] * 10001

        if amount == 0:
            return 0

        for i in range(1, amount+1):
            for j in coins:
                if i == j:
                    dp[j] = 1
                elif i > j:
                    dp[i] = min(dp[i], dp[i-j] + 1)

        if dp[amount] == sys.maxsize:
            return -1

        return dp[amount]

File: JunnyJun/Coin change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
     
        int result=0;
        sort(coins.begin(), coins.end(), greater<int>());
        auto ptr = coins.begin();
        while(amount && ptr!=coins.end()){
            if(amount >= (*ptr)){
                result += (amount/(*ptr));
                amount %= (*ptr);
            }
            ptr++;
        }
        return amount ? -1 : result;
    }
};